### PR TITLE
Refactor tests

### PR DESCRIPTION
**Pull request by quixotically**
## Description
Refactored tests into their own dir to clean up the code and to enable easy browser testing. Also fixed broken tests, and made other nit fixes.

## How Has This Been Tested?

See the changes I made to the README!

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: site/README.md [hunk 1]
```diff
@@ -15,6 +15,27 @@ Open http://localhost:8080/
 
 Anytime you change the contents, just refresh the page and it's going to be updated.
 
+# Run tests
+
+To run tests in the browser, do
+
+```
+npm run test:server
+Open http://localhost:8081/test
```
Comment (KamranAsif) [inline site/README.md]: nit: move this out of the code block

File: site/README.md [hunk 2]
```diff
@@ -15,6 +15,27 @@ Open http://localhost:8080/
 
 Anytime you change the contents, just refresh the page and it's going to be updated.
 
+# Run tests
+
+To run tests in the browser, do
```
Comment (KamranAsif) [inline site/README.md]: s/do/call

File: site/README.md [hunk 3]
```diff
@@ -15,6 +15,27 @@ Open http://localhost:8080/
 
 Anytime you change the contents, just refresh the page and it's going to be updated.
 
+# Run tests
```
Comment (KamranAsif) [inline site/README.md]: Lets move this whole section to CONTRIBUTING.md

File: src/FixedDataTableBufferedRows.js [hunk 4]
```diff
@@ -50,7 +50,6 @@ var FixedDataTableBufferedRows = React.createClass({
   },
 
   componentDidMount() {
-    setTimeout(this._updateBuffer, 1000);
```
Comment (quixotically) [inline src/FixedDataTableBufferedRows.js]: I'm hoping this wasn't needed, as _updateBuffer was not defined. Need some input here.

Before I removed this, >npm run test failed most FDTRoot-test tests with TypeError: "callback" argument must be a function, whereas they all passed with >npm run test:server. This is the difference I am referring to in my README.md changes.

Any idea what is going on there, or why we are seeing this difference?

Comment (KamranAsif) [inline src/FixedDataTableBufferedRows.js]: Why was this removed?

File: test/reducers/columnStateHelper-test.js [hunk 5]
```diff
@@ -180,15 +180,23 @@ describe('columnStateHelper', function() {
         isColumnResizing: true,
       };
 
+      const fakeElementTemplates = {
+        cell: [],
+        footer: [],
+        groupHeader: [],
+        header: []
+      };
```
Comment (quixotically) [inline test/reducers/columnStateHelper-test.js]: Is this ok fake data?

File: src/FixedDataTableBufferedRows.js (hunk 4 above)
Comment (quixotically) [inline src/FixedDataTableBufferedRows.js]: See my comment above

File: site/README.md (hunk 3 above)
Comment (quixotically) [inline site/README.md]: Done

File: site/README.md (hunk 2 above)
Comment (quixotically) [inline site/README.md]: done

File: site/README.md (hunk 1 above)
Comment (quixotically) [inline site/README.md]: done

File: site/README.md (hunk 2 above)
Comment (quixotically) [inline site/README.md]: actually, how about run? I think that's a better term for CL commands.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -28,5 +28,25 @@ We use GitHub issues to track public bugs. Please ensure your description is cle
...
+required! You can even click a specific section to only run those tests.
+
+To run tests in the CLI, run
+
+```
+npm run test
+```
+
+This will fail on some errors that npm run test:server won't, so it's recommended that you run this command after you've
+finished developing on tests with npm run test:server, if you chose that method.
```
Comment (wcjordan) [inline CONTRIBUTING.md]: Did you ever get a good idea of why this fails some cases where test:server won't?

File: src/FixedDataTableBufferedRows.js (hunk 4 above)
Comment (wcjordan) [inline src/FixedDataTableBufferedRows.js]: Hmm it looks like the method does exist in the file.   Rather than remove this, I think we need to figure out why it's undefined in this method call.  Maybe something bad with how the this object is bound?

File: test/reducers/columnStateHelper-test.js (hunk 5 above)
Comment (wcjordan) [inline test/reducers/columnStateHelper-test.js]: Ya that's fine.  It would be more ideal if we added some things here, but since these hold React elements they're harder to fake.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -28,5 +28,25 @@ We use GitHub issues to track public bugs. Please ensure your description is cle
 * "Attractive"
 * Do not use the optional parameters of `setTimeout` and `setInterval`
 
+## Testing
```
Comment (wcjordan) [inline CONTRIBUTING.md]: Nice, thanks for adding this section!

File: CONTRIBUTING.md (hunk 6 above)
Comment (KamranAsif) [inline CONTRIBUTING.md]: +1 this looks like a big issue. Do you have any examples of which ones fail this?

File: src/FixedDataTableBufferedRows.js (hunk 4 above)
Comment (KamranAsif) [inline src/FixedDataTableBufferedRows.js]: It might be related to how the test is creating this..can you debug the callstack to ComponentDidMount?

Comment (quixotically) [inline src/FixedDataTableBufferedRows.js]: It doesn't exist - if you're looking at https://github.com/schrodinger/fixed-data-table-2/search?utf8=%E2%9C%93&q=_updateBuffer&type= , you can see it was last indexed almost a month ago. This was the only usage left, and also even if the indexing was correct, you can see that _updateBuffer was not defined anywhere.

File: test/reducers/columnStateHelper-test.js (hunk 5 above)
Comment (quixotically) [inline test/reducers/columnStateHelper-test.js]: K yeah. Thanks!

File: src/FixedDataTableBufferedRows.js (hunk 4 above)
Comment (KamranAsif) [inline src/FixedDataTableBufferedRows.js]: This looks like my work. I probably deleted the logic and moved it to the reducer but forgot to clean this up.

this lgtm

File: CONTRIBUTING.md (hunk 6 above)
Comment (quixotically) [inline CONTRIBUTING.md]: https://github.com/schrodinger/fixed-data-table-2/pull/167#discussion_r119945925: "Before I removed this, >npm run test failed most FDTRoot-test tests with TypeError: "callback" argument must be a function, whereas they all passed with >npm run test:server"

That's all I got. I spent time investigating last week but to no avail, unfortunately. All I can think of is that the former runs mocha-webpack, while the latter runs webpack-dev-server, so maybe there's a difference in mocha-webpack? Not sure. I'll look into it a bit more.

File: CONTRIBUTING.md (hunk 7 above)
Comment (quixotically) [inline CONTRIBUTING.md]: :)

File: CONTRIBUTING.md (hunk 6 above)
Comment (quixotically) [inline CONTRIBUTING.md]: Here are the script differences, so you can see what I'm talking about. package.json scripts:
```
    "test": "mocha-webpack --webpack-config webpack.config-test.js \"test/**/*-test.js\" --require build_helpers/test-globals.js",
    "test:watch": "mocha-webpack --webpack-config webpack.config-test.js --watch \"test/**/*-test.js\" --require build_helpers/test-globals.js",
    "test:server": "webpack-dev-server --config webpack.config-test.js --hot --inline"
```

File: src/FixedDataTableBufferedRows.js (hunk 4 above)
Comment (quixotically) [inline src/FixedDataTableBufferedRows.js]: Ok cool - thanks for looking into it!

File: CONTRIBUTING.md (hunk 6 above)
Comment (quixotically) [inline CONTRIBUTING.md]: You know what, it's possible that the issue is that webpack-dev-server is not running in strict mode. Continuing to investigate.

Comment (quixotically) [inline CONTRIBUTING.md]: I've researched all I can, but still don't have a definite answer. My latest theory is that it's actually the difference between running in the browser and running in node.js. One observation is that there are new and different errors depending on node version, as can be seen in the travis results for my branch: https://travis-ci.org/schrodinger/fixed-data-table-2/builds/238907717?utm_source=github_status&utm_medium=notification.

But I can't be sure. Sorry!

Comment (KamranAsif) [inline CONTRIBUTING.md]: I'm 90% sure this is because of strict mode.
If you add 'use strict' to this top of that file and re-run; you should see identical errors.  
If thats the case, please file a ticket to enable strict mode in our tests

Comment (quixotically) [inline CONTRIBUTING.md]: I've tried and that doesn't happen, and that's why it's hard to pin down.

Comment (quixotically) [inline CONTRIBUTING.md]: It's almost as if something is stripping out 'use strict'.

Comment (quixotically) [inline CONTRIBUTING.md]: Oh but yeah when I said the difference was in node, I didn't mean that the issue was not use strict - I agree that that's what it looks like. What I meant was that maybe node handles use strict differently between versions. Once again though, I don't know enough to say for sure.

**Comment by wcjordan** [comment]
@KamranAsif  @quixotically  anything blocking merge here?  

Also what follow up work exists from investigating the strict mode / node issues

**Comment by quixotically** [comment]
Nope, nothing blocking merge afaik.

The follow up work is figuring out why >npm run test produces errors that >npm run test:server does not. My investigation left me with thinking strict mode and/or node was causing this difference - that's as far as I got. See this thread https://github.com/schrodinger/fixed-data-table-2/pull/167/files/df458e26ca43d5dfa3084cf7755305013804d012#r120031071 for more details.
